tests/test_vcs.py

```diff
@@ -143,3 +50,152 @@ def unknown_repo_type_url(request):
         vcs.identify_repo(unknown_repo_type_url)
+
+
+def test_prompt_should_ask_and_rm_repo_dir(mocker, tmpdir):
+    mock_read_user = mocker.patch(
+        'cookiecutter.vcs.read_user_yes_no',
+        return_value=True,
+        autospec=True
+    )
+    repo_dir = tmpdir.mkdir('repo')
+
+    vcs.prompt_and_delete_repo(str(repo_dir))
+
+    assert mock_read_user.called
+    assert not repo_dir.exists()
+
+
+def test_prompt_should_ask_and_keep_repo_dir(mocker, tmpdir):
+    mock_read_user = mocker.patch(
+        'cookiecutter.vcs.read_user_yes_no',
+        return_value=False,
+        autospec=True
+    )
+    repo_dir = tmpdir.mkdir('repo')
+
+    with pytest.raises(SystemExit):
+        vcs.prompt_and_delete_repo(str(repo_dir))
+
+    assert mock_read_user.called
+    assert repo_dir.exists()
+
+
+def test_prompt_should_not_ask_if_no_input_and_rm_repo_dir(mocker, tmpdir):
+    mock_read_user = mocker.patch(
+        'cookiecutter.vcs.read_user_yes_no',
+        return_value=True,
+        autospec=True
+    )
+    repo_dir = tmpdir.mkdir('repo')
+
+    vcs.prompt_and_delete_repo(str(repo_dir), no_input=True)
+
+    assert not mock_read_user.called
+    assert not repo_dir.exists()
+
+
+@pytest.fixture
+def clone_dir(tmpdir):
+    return str(tmpdir.mkdir('clone_dir'))
+
+
+def test_clone_should_raise_if_vcs_not_installed(mocker, clone_dir):
+    mocker.patch(
+        'cookiecutter.vcs.which',
+        autospec=True,
+        return_value=''
+    )
+
+    repo_url = 'https://github.com/pytest-dev/cookiecutter-pytest-plugin.git'
+
+    with pytest.raises(exceptions.VCSNotInstalled):
+        vcs.clone(repo_url, clone_to_dir=clone_dir)
+
+
+@pytest.mark.parametrize('repo_type, repo_url, repo_name', [
+    ('git', 'https://github.com/hello/world.git', 'world'),
+    ('hg', 'https://bitbucket.org/foo/bar', 'bar'),
+])
+def test_clone_should_invoke_git(
+        mocker, clone_dir, repo_type, repo_url, repo_name):
+    mocker.patch(
+        'cookiecutter.vcs.is_vcs_installed',
+        autospec=True,
+        return_value=True
+    )
+
+    mock_subprocess = mocker.patch(
+        'cookiecutter.vcs.subprocess.check_call',
+        autospec=True,
+    )
+    expected_repo_dir = os.path.normpath(os.path.join(clone_dir, repo_name))
+
+    branch = 'foobar'
+
+    repo_dir = vcs.clone(
+        repo_url,
+        checkout=branch,
+        clone_to_dir=clone_dir,
+        no_input=True
+    )
+
+    assert repo_dir == expected_repo_dir
+
+    mock_subprocess.assert_any_call(
+        [repo_type, 'clone', repo_url], cwd=clone_dir
+    )
+    mock_subprocess.assert_any_call(
+        [repo_type, 'checkout', branch], cwd=expected_repo_dir
+    )
+
+
+def test_clone_should_abort_if_user_does_not_want_to_reclone(mocker, tmpdir):
+    mocker.patch(
+        'cookiecutter.vcs.is_vcs_installed',
+        autospec=True,
+        return_value=True
+    )
+    mocker.patch(
+        'cookiecutter.vcs.read_user_yes_no',
```

Comment:
To make this more of a unitlike unit test, patch `cookiecutter.vcs.prompt_and_delete_repo` instead.